client/src/main/java/hudson/plugins/swarm/Client.java

```diff
@@ -285,4 +298,26 @@ private static boolean isDefaultOption(String key, Object value, CmdLineParser d
         }
         return false;
     }
+
+
+    public static class ShutdownHook extends Thread {
+        private SwarmClient client;
+        private Candidate target;
+
+        public ShutdownHook(SwarmClient client, Candidate target) {
+            this.client = client;
+            this.target = target;
+        }
+
+        @Override
+        public void run() {
+            logger.warning("Client shutdown initiated");
+            try {
+                client.takeSlaveOfflineAndWait(target, "JVM is shutting down");
+            } catch (Exception e) {
+                logger.log(Level.SEVERE, "Unable to perform graceful slave shutdown: ", e);
```

Comment:
What would happen if you started a shutdown operation (e.g., by sending a `SIGINT` with Ctrl-C or a `SIGTERM` with `kill(1)`) and then sent _another_ signal (e.g., another `SIGINT` by pressing Ctrl-C again) while we were in the middle of `takeSlaveOfflineAndWait()`? I think the expected behavior is that we would stop immediately without continuing to wait in `takeSlaveOfflineAndWait()`. I think that your current code probably already handles this in that an `InterruptedException` would be thrown (and you are catching `Exception` here). It might be worth validating this with a manual test. Even if the code works as expected, might it be worth printing an additional message in this case (e.g., "[terminating with extreme prejudice](https://en.wiktionary.org/wiki/terminate_with_extreme_prejudice)")?